### ISSUE TITLE
NoSuchMethodError on unobfuscated APK

### ISSUE THREAD

**Issue by cweggler** [appium/appium-espresso-driver#679]
**Details And Error**
Currently our Automation team uses Appium with UIAutomator2 but we are looking to make the switch to the espresso-driver to help with testing Android Recyclerviews.

On our tests that worked on the UIAutomator2 we are now seeing this error using the espresso-driver:
```
Proxying [POST /session] to [POST http://localhost:8200/session] with body: {"capabilities":{"firstMatch":[{"platform":"LINUX","webStorageEnabled":false,"takesScreenshot":true,"javascriptEnabled":true,"databaseEnabled":false,"networkConnectionEnabled":true,"locationContextEnabled":false,"warnings":{},"desired":{"platformName":"android","app":"/Users/rockwig/Downloads/app-debug.apk","autoGrantPermissions":true,"automationName":"espresso","deviceName":"emulator-5554","newCommandTimeout":10000,"resetKeyboard":true,"systemPort":8200,"unicodeKeyboard":true},"platformName":"android","app":"/Users/rockwig/Downloads/app-debug.apk","autoGrantPermissions":true,"automationName":"espresso","deviceName":"emulator-5554","newCommandTimeout":10000,"resetKeyboard":true,"systemPort":8200,"unicodeKeyboard":true,"deviceUDID":"emulator-5554","appPackage":"com.ourfamilywizard","appWaitPackage":"com.ourfamilywizard","appActivity":"com.ourfamilywizard.login.activity.LoginActivity","appWaitActivity":"com.ourfamilywizard.login.activity.LoginActivity"}],"alwaysMatch":{}}}
[debug] [Espresso] [Instrumentation] Process crashed while executing startEspressoServer(io.appium.espressoserver.EspressoServerRunnerTest):
[debug] [Espresso] java.lang.NoSuchMethodError: No direct method <init>(Ljava/lang/Class;Ljava/lang/String;Ljava/lang/String;I)V in class Lkotlin/jvm/internal/PropertyReference1Impl; or its super classes (declaration of 'kotlin.jvm.internal.PropertyReference1Impl' appears in /data/app/~~D3o6gWwUAGVZL4YA0w1cHQ==/io.appium.espressoserver.test-ELPxSt8tB_Ojzmg9MoOxlw==/base.apk)
[debug] [Espresso] 	at com.ourfamilywizard.login.activity.LoginActivity.<clinit>(Unknown Source:12)
[debug] [Espresso] 	at java.lang.Class.newInstance(Native Method)
[debug] [Espresso] 	at android.app.AppComponentFactory.instantiateActivity(AppComponentFactory.java:95)
[debug] [Espresso] 	at androidx.core.app.CoreComponentFactory.instantiateActivity(CoreComponentFactory.java:45)
[debug] [Espresso] 	at android.app.Instrumentation.newActivity(Instrumentation.java:1253)
[debug] [Espresso] 	at androidx.test.runner.MonitoringInstrumentation.newActivity(MonitoringInstrumentation.java:798)
[debug] [Espresso] 	at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:3353)
[debug] [Espresso] 	at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:3601)
[debug] [Espresso] 	at android.app.servertransaction.LaunchActivityItem.execute(LaunchActivityItem.java:85)
[debug] [Espresso] 	at android.app.servertransaction.TransactionExecutor.executeCallbacks(TransactionExecutor.java:135)
[debug] [Espresso] 	at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:95)
[debug] [Espresso] 	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2066)
[debug] [Espresso] 	at android.os.Handler.dispatchMessage(Handler.java:106)
[debug] [Espresso] 	at android.os.Looper.loop(Looper.java:223)
[debug] [Espresso] 	at android.app.ActivityThread.main(ActivityThread.java:7656)
[debug] [Espresso] 	at java.lang.reflect.Method.invoke(Native Method)
[debug] [Espresso] 	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:592)
[debug] [Espresso] 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:947)
```

**Background/ What we already tried:**
We did see this previous issue [here](https://github.com/appium/appium-espresso-driver/issues/232) and tried running our test on our debug APK which we do not add obfuscation to but still got the same error. So this leads us to think that this issue is not related to Proguard.

Our app does have a fragment on the initial launch of the Login Activity and uses the Jetpack Navigation Library, if that helps.

**Comment by KazuCocoa** [comment]
I guess it could be Kotlin related issue.
https://github.com/appium/appium-espresso-driver#toolsversions
e.g. https://youtrack.jetbrains.com/issue/KT-37435

Current master uses 1.4.32 as kotlin by default, https://github.com/appium/appium-espresso-driver/blob/67cb813c135a3acc097dcb406464491008d421e0/espresso-server/build.gradle.kts#L5 , but previous one as 1.3.72.

If the kotlin version was different from the app under test, changing the kotlin version may help...
You also can try out latest `appium@beta` to try the kotlin version.

**Comment by cweggler** [comment]
@KazuCocoa - Thanks for your reply.

We tried creating another debug APK with the exact Kotlin version you mentioned: `"1.4.32"` and re-ran the test but still get the same error.

**Comment by cweggler** [comment]
Another idea- but could this be related to the app using a `FragmentFactory` to instantiate fragments. Would that lead to this error?

**Comment by KazuCocoa** [comment]
What version of Kotlin did your app use?

The cause is probably a method call in your LoginActivity referred a Java property, but the bundled kotlin did not have it.
https://github.com/appium/appium-espresso-driver/blob/master/espresso-server/app/build.gradle.kts#L64-L104 is the list of libraries espresso driver built. I guess kotlin-reflect or kotlin-stdlib versions could be related to, or this espresso driver should use the same kotlin version of yours.

```
[debug] [Espresso] java.lang.NoSuchMethodError: No direct method <init>(Ljava/lang/Class;Ljava/lang/String;Ljava/lang/String;I)V in class Lkotlin/jvm/internal/PropertyReference1Impl; or its super classes (declaration of 'kotlin.jvm.internal.PropertyReference1Impl' appears in /data/app/~~D3o6gWwUAGVZL4YA0w1cHQ==/io.appium.espressoserver.test-ELPxSt8tB_Ojzmg9MoOxlw==/base.apk)
[debug] [Espresso] 	at com.ourfamilywizard.login.activity.LoginActivity.<clinit>(Unknown Source:12)
[debug] [Espresso] 	at java.lang.Class.newInstance(Native Method)
```

**Comment by cweggler** [comment]
I used Kotlin version `1.4.32` on our test and we did use the same versions for `kotlin-reflect` and `kotlin-stdlib`. But I'm starting to think this error is because of our Dagger dependency injection setup in the Android app itself. Dependency Injection hasn't been set up for tests. When we try to run a simple Espresso test locally (not using Appium) we get the same error, but more information on why-
```
Caused by: java.lang.NoSuchMethodException: <init> []
	at java.lang.Class.getConstructor0(Class.java:2320)
	at java.lang.Class.getConstructor(Class.java:1725)
	at androidx.fragment.app.FragmentFactory.instantiate(FragmentFactory.java:121)
	... 15 more
```
